retrievalmarket/network/libp2p_impl_test.go

```diff
@@ -0,0 +1,516 @@
+	if tr.queryStreamHandler != nil {
+		tr.queryStreamHandler(s)
+	}
+}
+
+func TestQueryStreamSendReceiveQuery(t *testing.T) {
+	ctx := context.Background()
+	td := shared_testutil.NewLibp2pTestData(ctx, t)
+
+	fromNetwork := network.NewFromLibp2pHost(td.Host1)
+	toNetwork := network.NewFromLibp2pHost(td.Host2)
+	toHost := td.Host2.ID()
+
+	// host1 gets no-op receiver
+	tr := &testReceiver{t: t}
+	require.NoError(t, fromNetwork.SetDelegate(tr))
+
+	// host2 gets receiver
+	qchan := make(chan retrievalmarket.Query)
+	tr2 := &testReceiver{t: t, queryStreamHandler: func(s network.RetrievalQueryStream) {
+		readq, err := s.ReadQuery()
+		require.NoError(t, err)
+		qchan <- readq
+	}}
+	require.NoError(t, toNetwork.SetDelegate(tr2))
+
+	// setup query stream host1 --> host 2
+	assertQueryReceived(ctx, t, fromNetwork, toHost, qchan)
+}
+
+func TestQueryStreamSendReceiveQueryResponse(t *testing.T) {
+	ctx := context.Background()
+	td := shared_testutil.NewLibp2pTestData(ctx, t)
+	fromNetwork := network.NewFromLibp2pHost(td.Host1)
+	toNetwork := network.NewFromLibp2pHost(td.Host2)
+	toHost := td.Host2.ID()
+
+	// host1 gets no-op receiver
+	tr := &testReceiver{t: t}
+	require.NoError(t, fromNetwork.SetDelegate(tr))
+
+	// host2 gets receiver
+	qchan := make(chan retrievalmarket.QueryResponse)
+	tr2 := &testReceiver{t: t, queryStreamHandler: func(s network.RetrievalQueryStream) {
+		q, err := s.ReadQueryResponse()
+		require.NoError(t, err)
+		qchan <- q
+	}}
+	require.NoError(t, toNetwork.SetDelegate(tr2))
+
+	assertQueryResponseReceived(ctx, t, fromNetwork, toHost, qchan)
+
+}
+
+func TestQueryStreamSendReceiveMultipleSuccessful(t *testing.T) {
+	// send query, read in handler, send response back, read response
+	ctxBg := context.Background()
+	td := shared_testutil.NewLibp2pTestData(ctxBg, t)
+	nw1 := network.NewFromLibp2pHost(td.Host1)
+	nw2 := network.NewFromLibp2pHost(td.Host2)
+	require.NoError(t, td.Host1.Connect(ctxBg, peer.AddrInfo{ID: td.Host2.ID()}))
+
+	// host2 gets a query and sends a response
+	qr := shared_testutil.MakeTestQueryResponse()
+	done := make(chan bool)
+	tr2 := &testReceiver{t: t, queryStreamHandler: func(s network.RetrievalQueryStream) {
+		_, err := s.ReadQuery()
+		require.NoError(t, err)
+
+		require.NoError(t, s.WriteQueryResponse(qr))
+		done <- true
+	}}
+	require.NoError(t, nw2.SetDelegate(tr2))
+
+	ctx, cancel := context.WithTimeout(ctxBg, 10*time.Second)
+	defer cancel()
+
+	qs, err := nw1.NewQueryStream(td.Host2.ID())
+	require.NoError(t, err)
+
+	testCid := testutil.GenerateCids(1)[0]
+
+	var resp retrievalmarket.QueryResponse
+	go require.NoError(t, qs.WriteQuery(retrievalmarket.Query{PieceCID: testCid.Bytes()}))
+	resp, err = qs.ReadQueryResponse()
+	require.NoError(t, err)
+
+	select {
+	case <-ctx.Done():
+		t.Error("response not received")
+	case <-done:
+	}
+
+	assert.Equal(t, qr, resp)
+}
+
+func TestQueryStreamSendReceiveOutOfOrderFails(t *testing.T) {
+	// send query, read response in handler - fails
```

Comment:
I had in mind just doing:
from: write query
to: read query response <- errors

Not the second step as well. It's fine that the test is actually testing more, but if it means you can get rid of the go routine I'd probably just do that.